### PR TITLE
imgbundler and CLI fixes

### PR DESCRIPTION
**Pull request by nhooyr**
See individual commits.

And please test @alixander

### CODE REVIEW

**Comment by nhooyr** [comment]
Example:

#### script

```d2
server: {
  shape: image
  icon: https://iconss.terrastruct.com/tech/022-server.svg
}

github: {
  shape: image
  icon: https://icon2.terrastruct.com/dev/github.svg
}

tri: {
  shape: image
  icon: /tmp/jingle.png
}

server -> github
```

#### watch terminal

image

#### watch render

Screenshot 2022-11-30 at 3 08 58 AM

#### prod terminal

Screenshot 2022-11-30 at 3 11 11 AM

#### prod render

Screenshot 2022-11-30 at 3 11 32 AM

#### thoughts

not a fan of the prod render, maybe we ought to include the error within the svg just like I was doing before.
Or at the very least replace the image with a placeholder.

**Comment by nhooyr** [comment]
I opened https://github.com/terrastruct/d2/issues/280

File: cmd/d2/watch.go [hunk 1]
```diff
@@ -345,19 +346,23 @@ func (w *watcher) compileLoop(ctx context.Context) error {
...
-		b, err := compile(ctx, w.ms, true, w.layoutPlugin, w.themeID, w.inputPath, w.outputPath, w.pw.Page)
+		b, _, err := compile(ctx, w.ms, w.layoutPlugin, w.themeID, w.inputPath, w.outputPath, w.bundle, w.pw.Page)
+		errs := ""
 		if err != nil {
-			err = fmt.Errorf("failed to %scompile: %w", recompiledPrefix, err)
-			w.ms.Log.Error.Print(err)
-			w.broadcast(&compileResult{
-				Err: err.Error(),
-			})
+			if len(b) > 0 {
```
Comment (alixander) [inline cmd/d2/watch.go:352]: can't you do the same in other place? (instead of written bool)

File: lib/imgbundler/imgbundler.go [hunk 2]
```diff
@@ -6,156 +6,202 @@ import (
...
-	var filtered [][][]byte
+// filterImageElements finds all image elements in imgs that are eligible
+// for bundling in the current context.
+func filterImageElements(imgs [][][]byte, isRemote bool) [][][]byte {
+	imgs2 := imgs[:0]
 	for _, img := range imgs {
-		u, err := url.Parse(string(img[1]))
+		href := string(img[1])
+
+		// Skip already bundled images.
```
Comment (alixander) [inline lib/imgbundler/imgbundler.go:64]: oh good catch

File: lib/imgbundler/imgbundler.go [hunk 3]
```diff
@@ -6,156 +6,202 @@ import (
...
-	return fmt.Sprintf("data:%s;base64,%s", mimeType, enc), nil
+	return buf, resp.Header.Get("Content-Type"), nil
 }
 
-func read(href string) (string, error) {
-	data, err := os.ReadFile(href)
-	if err != nil {
-		return "", err
+// sniffMimeType sniffs the mime type of href based on its file extension and contents.
+func sniffMimeType(href, buf []byte, isRemote bool) string {
```
Comment (alixander) [inline lib/imgbundler/imgbundler.go:192]: nice

File: lib/imgbundler/imgbundler.go [hunk 4]
```diff
@@ -6,156 +6,202 @@ import (
...
 
-				var data string
-				var err error
-				if isRemote {
-					data, err = fetch(ctx, href)
-				} else {
-					data, err = read(href)
+				bundledImage, err := worker(ctx, img[1], isRemote)
+				if err != nil {
+					ms.Log.Error.Printf("failed to bundle %s: %v", img[0], err)
```
Comment (alixander) [inline lib/imgbundler/imgbundler.go:108]: u want to log `img[1]`, the href

Screen Shot 2022-11-30 at 10 18 21 AM

**Review by alixander — CHANGES_REQUESTED** [review]
👏 

add in the release notes that the bundle flag now works. 

removing the conversion from `text/xml` to `image/svg+xml` would cause issues viewing in browser i think. it's why i did it in the first place: https://stackoverflow.com/questions/31755822/why-does-this-svg-work-being-viewed-raw-in-the-browser-but-not-in-a-webpage

**Comment by alixander** [comment]
for me, it doesn't show the broken img. 

Screen Shot 2022-11-30 at 10 17 08 AM

File: cmd/d2/main.go [hunk 5]
```diff
@@ -167,27 +169,26 @@ func run(ctx context.Context, ms *xmain.State) (err error) {
...
-	if *bundleFlag {
-		_ = 343
-	}
-
-	_, err = compile(ctx, ms, false, plugin, *themeFlag, inputPath, outputPath, pw.Page)
+	_, written, err := compile(ctx, ms, plugin, *themeFlag, inputPath, outputPath, *bundleFlag, pw.Page)
 	if err != nil {
-		return err
+		if written {
+			return fmt.Errorf("failed to compile (partial render written): %w", err)
```
[the remark this answers is not in the document]
Comment (ejulio-ts) [reply cmd/d2/main.go:175]: I guess that failing to compile shouldn't generate a partial output.
Maybe, naming of this function is odd as it compiles and renders..
If compilation fails, there is no way we can render.
If render fails, we might have been able to generate a partial output.

looking into `compile` it does way more stuff than just compiling, it is more to `run` as the main function of the CLI

File: cmd/d2/watch.go (hunk 1 above)
Comment (ejulio-ts) [reply cmd/d2/watch.go:352]: what is `b` here? 😂

Comment (alixander) [reply cmd/d2/watch.go:352]: bytes

Comment (ejulio-ts) [reply cmd/d2/watch.go:352]: it was more in the sense, inside this function it is called `svg` and the return type is `bytes` so there is a lot context lost in just `b`.

File: cmd/d2/main.go [hunk 6]
```diff
@@ -201,47 +202,46 @@ func compile(ctx context.Context, ms *xmain.State, isWatching bool, plugin d2plu
...
+		bundleErr = multierr.Combine(bundleErr, bundleErr2)
 	}
 
 	out := svg
 	if filepath.Ext(outputPath) == ".png" {
-		svg, err = imgbundler.InlineRemote(ctx, ms, svg)
-		if err != nil {
-			ms.Log.Error.Printf("missing/broken remote image(s), writing partial output: %v", err)
+		svg := svg
+		if !bundle {
```
[the remark this answers is not in the document]
Comment (nhooyr) [reply cmd/d2/main.go:227]: Not sure what you mean. if `!bundle` we ensure to bundle remote images. But if `bundle` then the remote imgaes have already been bundled.

File: cmd/d2/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nhooyr) [reply cmd/d2/main.go:175]: We already have a `run` that does a bunch of things. I'll concede it's unfortuante but I don't know what else to name it. `compile` is our shared compilation/writing output code between watch mode and oneshot (prod) mode.

@alixander is `failed to fully compile` ok? I prefer we keep the phrasing close.

File: cmd/d2/watch.go (hunk 1 above)
Comment (nhooyr) [reply cmd/d2/watch.go:352]: @alixander No you can get a partial svg even if the png/output failed to write.

@ejulio-ts yea fair enough i renamed thanks

File: lib/imgbundler/imgbundler.go (hunk 3 above)
Comment (nhooyr) [reply lib/imgbundler/imgbundler.go:192]: Yea the github image was being detected as text/plain for me lol.

**Comment by nhooyr** [comment]
@alixander what's your d2 text there?

**Comment by alixander** [comment]
```d2
test: {
  icon: https://no.com
}

test2: {
  icon: https://no.com
}

works: {
  icon: https://www.howtogeek.com/wp-content/uploads/2018/06/shutterstock_1006988770.png
}

test -> test2
```

File: cmd/d2/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (alixander) [reply cmd/d2/main.go:227]: yeah my b misunderstood

**Comment by nhooyr** [comment]
Weird I can confirm, don't see the image either. I don't know why in that particular example. Mine still works as expected.

Screenshot 2022-11-30 at 2 48 55 PM

Could it be unrelated to this PR? I'm not sure how any of my changes could have caused this.

**Comment by alixander** [comment]
yeah, maybe some SVG thing. we'll just note in that issue.

**Review by alixander — APPROVED** [review]

**Pull request merged**